### ISSUE TITLE
[distributed] Bringing minio server instance down/up leads to intermittent panic

### ISSUE THREAD

**Issue by krisis** [minio/minio#3111]
In a docker-compose setup of 4 nodes (1 disk each), when uploading objects of size 65MB (multipart) using minio-py SDK, one docker instance was brought down for a brief period of time.
At this point we saw panic traces in the docker instance that was brought down and back up again.
## Expected Behavior

No panics should be observed when nodes are brought down and back up.
## Current Behavior

```
minio3_1  | 2016/10/27 09:54:41 http: panic serving 172.19.0.1:46294: Unable to look for valid XL metadata content
minio3_1  | goroutine 1310692 [running]:
minio3_1  | net/http.(*conn).serve.func1(0xc42ac71200)
minio3_1  |     /usr/local/go/src/net/http/server.go:1491 +0x12a
minio3_1  | panic(0xa87d20, 0xc4d8e28110)
minio3_1  |     /usr/local/go/src/runtime/panic.go:458 +0x243
minio3_1  | github.com/minio/minio/cmd.pickValidXLMeta(0xc4df819500, 0x4, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
minio3_1  |     /go/src/github.com/minio/minio/cmd/xl-v1-metadata.go:204 +0x277
minio3_1  | github.com/minio/minio/cmd.xlObjects.PutObjectPart(0xc4203a4fc0, 0x4, 0x4, 0x2, 0x2, 0x2, 0x3, 0xc420305e00, 0xc4203a5000, 0x1, ...)
minio3_1  |     /go/src/github.com/minio/minio/cmd/xl-v1-multipart.go:491 +0xf3b
minio3_1  | github.com/minio/minio/cmd.(*xlObjects).PutObjectPart(0xc4203926e0, 0xc43c78d115, 0x10, 0xc43c78d126, 0x13, 0xc43c78d151, 0x24, 0xa, 0x500000, 0x7efc5a4b29f8, ...)
minio3_1  |     <autogenerated>:837 +0x14c
minio3_1  | github.com/minio/minio/cmd.objectAPIHandlers.PutObjectPartHandler(0xbf1be8, 0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a1e0)
minio3_1  |     /go/src/github.com/minio/minio/cmd/object-handlers.go:688 +0x6b1
minio3_1  | github.com/minio/minio/cmd.(objectAPIHandlers).PutObjectPartHandler-fm(0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a1e0)
minio3_1  |     /go/src/github.com/minio/minio/cmd/api-router.go:47 +0x4e
minio3_1  | net/http.HandlerFunc.ServeHTTP(0xc4203b1750, 0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a1e0)
minio3_1  |     /usr/local/go/src/net/http/server.go:1726 +0x44
minio3_1  | github.com/minio/minio/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc4203930e0, 0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a1e0)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:107 +0x10d
minio3_1  | github.com/minio/minio/vendor/github.com/urfave/negroni.Wrap.func1(0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a000, 0xc45e925b00)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/urfave/negroni/negroni.go:41 +0x4d
minio3_1  | github.com/minio/minio/vendor/github.com/urfave/negroni.HandlerFunc.ServeHTTP(0xc4203fb640, 0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a000, 0xc45e925b00)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/urfave/negroni/negroni.go:24 +0x4e
minio3_1  | github.com/minio/minio/vendor/github.com/urfave/negroni.middleware.ServeHTTP(0xef8e40, 0xc4203fb640, 0xc4203fb6a0, 0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a000)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/urfave/negroni/negroni.go:33 +0xa7
minio3_1  | github.com/minio/minio/vendor/github.com/urfave/negroni.(middleware).ServeHTTP-fm(0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a000)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/urfave/negroni/negroni.go:33 +0x60
minio3_1  | github.com/minio/minio/vendor/github.com/urfave/negroni.Wrap.func1(0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a000, 0xc45e925a40)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/urfave/negroni/negroni.go:42 +0x74
minio3_1  | github.com/minio/minio/vendor/github.com/urfave/negroni.HandlerFunc.ServeHTTP(0xc4203fb620, 0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a000, 0xc45e925a40)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/urfave/negroni/negroni.go:24 +0x4e
minio3_1  | github.com/minio/minio/vendor/github.com/urfave/negroni.middleware.ServeHTTP(0xef8e40, 0xc4203fb620, 0xc4203fb680, 0x7efc5a4b29c0, 0xc42cc76b40, 0xc4ec07a000)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/urfave/negroni/negroni.go:33 +0xa7
minio3_1  | github.com/minio/minio/vendor/github.com/urfave/negroni.(*Negroni).ServeHTTP(0xc4203e7800, 0xefd340, 0xc420586000, 0xc4ec07a000)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/urfave/negroni/negroni.go:73 +0xdd
minio3_1  | github.com/minio/minio/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc4202ca8c0, 0xefd340, 0xc420586000, 0xc4ec07a000)
minio3_1  |     /go/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:107 +0x10d
minio3_1  | github.com/minio/minio/cmd.requestSizeLimitHandler.ServeHTTP(0xef65c0, 0xc4202ca8c0, 0x180000000, 0xefd340, 0xc420586000, 0xc490fd7e00)

```
## Steps to Reproduce (for bugs)

1. `docker-compose up` from minio server source repo.
2. `mc mb <alias>/kp-test-minio-py`
3. `seq 5  | xargs -n 1 -P 32 -I{} -t python small-objects-test.py --count 5`
4. Use [small-objects-test.py](https://gist.github.com/krisis/031ef243a1033156a7671d2b76872f01)
5. While the upload is ongoing, bring one of the docker instance down.
   say `docker stop minio_minio3_1`.
   ## Context
   
   
## Your Environment

- Version used:

```
Version: 2016-10-27T03:13:33Z
Release-Tag: DEVELOPMENT.2016-10-27T03-13-33Z
Commit-ID: 30dc11a931010cdca52dddddb7b06fafbe0db8b2
```
- Environment name and version (e.g. nginx 1.9.1): 
- Server type and version: `docker-compose` based

**Comment by vadmeste** [comment]
I am taking this

**Comment by vadmeste** [comment]
A manual panic()  call in the code was executed, but wasn't able to reproduce it.

**Comment by vadmeste** [comment]
> A manual panic() call in the code was executed, but wasn't able to reproduce it.

wouldn't it be better if we just return a meaningful error here instead of a panic ?

**Comment by harshavardhana** [comment]
> wouldn't it be better if we just return a meaningful error here instead of a panic ?

It is better but we need to investigate why this happens first.

**Comment by harshavardhana** [comment]
This might not happen with current master @krisis @vadmeste we will have to reproduce this once more.

**Comment by GarimaKapoor** [comment]
@balamurugana can you see if this is still valid?

**Comment by krisis** [comment]
I can take this up.

**Comment by GarimaKapoor** [comment]
@krisis what is the timeline for the fix of this issue.

**Comment by krisis** [comment]
I am looking into it alongside https://github.com/minio/minio/issues/3274. I expect to have more findings earlier than EOD Nov 19th.

**Comment by krisis** [comment]
## Test description

The python script used minio1 to perform the multipart upload. minio2 was brought down and back up while many concurrent multipart uploads were ongoing.

[Output snippets of backtrace, backend structure](https://gist.github.com/krisis/2e72c7ba7585768ae6662e805b9aa6f6) used in the following analysis.

## Analysis

The backend layout in the above gist for `my-boring-object-3` indicates that a CompleteMultipartUpload happened on upload id `cf06534c-bcee-405c-976d-b22a3bdb71a7` on `minio1`, `minio3` and `minio4` before the PutObjectPart API request (which panic'd). We know this from the missing `uploads.json` and `xl.json` under the object directory. `minio2` must have come up after the part (i.e, `part.10`) was written on the rest. We also know from the backend layout that `minio2` didn't witness the same CompleteMultipartUpload request. `xl.isUploadExists` returns true if the part has xl.json in one of the disks. In this case from `minio2`. This is _incorrect_. From the layout, we can see that quorum number of disks agree that the upload id doesn't exist. The panic occurred since there is no valid `xlMetaV1` value from `onlineDisks`, which consists of nodes `minio{1,3,4}`. This is a direct consequence of disregarding quorum and picking any one node that has `xl.json`.

## Solution

We should add quorum check to `xl.isUploadIDExists` function to avoid this panic. Additionally, we should error out PutObjectPart, with a log message when this condition occurs, instead of a panic.

**Comment by krisis** [comment]
@harshavardhana @krishnasrinivas @donatello https://github.com/minio/minio/pull/3202 will fix this issue too. Do the objections in the PR apply to this case as well?

**Comment by krisis** [comment]
As per the closing comment in #3202 if we 'heal' `uploads.json` for all pending multipart uploads during startup of a minio server instance, this issue would also be addressed.

**Comment by harshavardhana** [comment]
> We should add quorum check to xl.isUploadIDExists function to avoid this panic. Additionally, we should error out PutObjectPart, with a log message when this condition occurs, instead of a panic.

We decided to heal the `uploads.json` rather than bringing quorum checks as stated in #3202, i am working on it. 

Here on the other hand the panic should be still addressed by not panicking but cleanly returning the values upwards through the handlers back to the client.

**Comment by krisis** [comment]
OK, I shall send a PR to remove the panic and return an error instead, failing the PutObjecPart API request.

**Comment by harshavardhana** [comment]
Fixed by #3277
